nntrainer/tensor/manager.h

```diff
@@ -276,57 +295,28 @@ class Manager {
    * @brief Deallocate memory for all the managed tensors
    */
   void deallocateTensors(bool dealloc_weights = false) {
```

Comment:
Could you specify the semantics of `tensors`?
it seems to state one of three from time to time.
1. gradient + weights + inouts + derivatives
2. gradient + inouts + derivatives
3. inouts + derivatives.

Reply:
I will rename the function to `deallocateAllTensors()`.
Tensors will consistently refer to `gradient + inouts + derivatives`.
And all tensors will include weights as well.

Reply:
Actually, deallocateTensors() is supposed to give same abstraction as allocateTensors() and initializeTensors().
which is meant for `gradient + inouts + derivatives`.
deallocateTensors() provides an additional option to deallocate weights as well (but does not by default).
allocateTensors() ensures that weights are allocated internally (intended to allocate the trio only).

Please suggest alternative if this is confusing.

Reply:
My concern is about the invariants seems,

InitializeTensors, might not allocates gradients (if it is not trainable).
```cpp
void func1() {
  initializeTensors(false); // initailzes weight + inout + deriv
  initializeTensors(true); // this doesn't initialize grad although it should(?) or fail because already initialized?
}

void func2() {
  initilizeTensors(false); // initialize weight + inout + deriv
  allocateTensors(); // allocates weight + inout + deriv + grad (it shouldn't allocate grad) 
  deallocateTensors(); // deallocates grad (it shouldn't deallocate grad) 
}

void func3() {
  allocateTensors(); // if it is not initialized, should this be hard error? if initialized before, is grad initialized or not?
}

void func4() {
  deallocateTensors(); // if not allocated before, Should this be an hard error?
}
```

Reply:
check #981 once, if it does what you need.

> InitializeTensors, might not allocates gradients (if it is not trainable).

`InitializeTensors` will never allocate gradient, only initialize. But, if `initializeTensors` didn't initialize gradients, then gradients wont be initialized. `funcTensors` provides an abstraction which applies processing the `func` on all 3 - `gradients, inouts and derivatives`, as if the same `func` was applied on all 3 individually. In case of not trainable, `initializeGradients` would have not done anything. 
But that is now handled by `initializeTensors` and `initializeGradients` just initialize gradients if asked to. Decision making has moved to `funcTensors`.

Reply:
I checked #981, it's orthogonal to this.

I'll get to you in person to explain.

Reply:
Discussed offline, this will be treated with a separate PR.

Reply:
This will be handled in or over #981.